File: Second Year/2nd Semester/Medii de Proiectare si Programare/C#/Laborator_2/repository/ResultDBRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Runtime.CompilerServices;
using Laborator_1.domain;
using log4net;

namespace Laborator_1.repository
{
    public class ResultDBRepository:ResultRepository
    {
        private static readonly ILog log = LogManager.GetLogger("ResultDBRepository");
        private IDictionary<string, string> props;
        private RefereeRepository refereeRepository;
        private ParticipantRepository participantRepository;
        
        public ResultDBRepository(IDictionary<string, string> props, RefereeDBRepository refereeDbRepository, ParticipantDBRepository participantDbRepository)
        {
            log.Info("Creating ResultDBRepository");
            this.props = props;
            refereeRepository = refereeDbRepository;
            participantRepository = participantDbRepository;
        }
        
        public Result FindOne(int id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection conn = DBUtils.getConnection(props);
            using (var command = conn.CreateCommand())
            {
                command.CommandText = "select id,id_referee,id_participant,points,activity from Results where id=@id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                using (var dataR = command.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idV = dataR.GetInt32(0);
                        int idR = dataR.GetInt32(1);
                        Referee referee = refereeRepository.FindOne(idR);
                        int idP = dataR.GetInt32(2);
                        Participant participant = participantRepository.FindOne(idP);
                        int points = dataR.GetInt32(3);
                        string activity = dataR.GetString(4);
                        Result result = new Result(idV, participant, referee, points, activity);
                        log.InfoFormat("Existing findOne with value {0}", result);
                        return result;
                    }
                }
            }
            log.InfoFormat("Existing findOne with value {0}", null);
            return null;
        }

        public IEnumerable<Result> FindAll()
        { 
            IDbConnection conn = DBUtils.getConnection(props);
            IList<Result> results = new List<Result>();
            using (var command = conn.CreateCommand())
            {
                command.CommandText = "select id,id_referee,id_participant,points,activity from Results";

                using (var dataR = command.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idV = dataR.GetInt32(0);
                        int idR = dataR.GetInt32(1);
                        Referee referee = refereeRepository.FindOne(idR);
                        int idP = dataR.GetInt32(2);
                        Participant participant = participantRepository.FindOne(idP);
                        int points = dataR.GetInt32(3);
                        string activity = dataR.GetString(4);
                        Result result = new Result(idV, participant, referee, points, activity);
                        results.Add(result);
                    }
                }
            }

            return results;
        }

        public void Save(Result entity)
        {
            var conn = DBUtils.getConnection(props);
            using (var command = conn.CreateCommand())
            {
                command.CommandText =
                    "insert into Results values(@id, @id_ref, @id_part, @points, @activity)";
                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = entity.Id;
                command.Parameters.Add(paramId);
                
                var paramIdR = command.CreateParameter();
                paramIdR.ParameterName = "@id_ref";
                paramIdR.Value = entity.Referee.Id;
                command.Parameters.Add(paramIdR);
                
                var paramIdP = command.CreateParameter();
                paramIdP.ParameterName = "@id_part";
                paramIdP.Value = entity.Participant.Id;
                command.Parameters.Add(paramIdP);

                var paramActivity = command.CreateParameter();
                paramActivity.ParameterName = "@activity";
                paramActivity.Value = entity.Activity;
                command.Parameters.Add(paramActivity);

                var paramPoints = command.CreateParameter();
                paramPoints.ParameterName = "@points";
                paramPoints.Value = entity.Points;
                command.Parameters.Add(paramPoints);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No result added!");
                
                // Actualizarea punctajului participantului
                int new_points = entity.Participant.Points + entity.Points;
                Participant new_participant = new Participant(entity.Participant.Id,entity.Participant.First_Name,entity.Participant.Last_Name,new_points);
                participantRepository.Update(new_participant);
            }
        }

        public void Delete(int id)
        {
            IDbConnection conn = DBUtils.getConnection(props);
            using (var command = conn.CreateCommand())
            {
                command.CommandText = "delete from Results where id=@id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);
                var dataR = command.ExecuteNonQuery();
                if (dataR == 0)
                    throw new RepositoryException("No referee deleted!");
                Result entity = FindOne(id);
                int new_points = entity.Participant.Points -  entity.Points;
                Participant new_participant = new Participant(entity.Participant.Id,entity.Participant.First_Name,entity.Participant.Last_Name,new_points);
                this.participantRepository.Update(new_participant);
            }
        }

        public void Update(Result entity)
        {
            var conn = DBUtils.getConnection(props);
            int old_points = this.participantRepository.FindOne(entity.Participant.Id).Points; // puntajul vechi
            using (var command = conn.CreateCommand())
            {
                command.CommandText =
                    "update Results set id_referee = @id_ref, id_participant = @id_part, points = @points, activity = @activity where id = @id)";
                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = entity.Id;
                command.Parameters.Add(paramId);

                var paramR = command.CreateParameter();
                paramR.ParameterName = "@id_ref";
                paramR.Value = entity.Referee.Id;
                command.Parameters.Add(paramR);

                var paramP = command.CreateParameter();
                paramP.ParameterName = "@id_part";
                paramP.Value = entity.Participant.Id;
                command.Parameters.Add(paramP);
                
                var paramPoints = command.CreateParameter();
                paramPoints.ParameterName = "@points";
                paramPoints.Value = entity.Points;
                command.Parameters.Add(paramPoints);

                var paramActivity = command.CreateParameter();
                paramActivity.ParameterName = "@activity";
                paramActivity.Value = entity.Activity;
                command.Parameters.Add(paramActivity);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No result updated!");
                
                //actualizam punctajul participantului
                int new_points = entity.Participant.Points - old_points + entity.Points;
                Participant new_participant = new Participant(entity.Participant.Id,entity.Participant.First_Name,entity.Participant.Last_Name,new_points);
                this.participantRepository.Update(new_participant);
            }
        }

        public List<Participant> getEvaluatedParticipants(Referee referee)
        {
            log.InfoFormat("Entering getEvaluatedParticipants with value {0}", referee);
            IDbConnection conn = DBUtils.getConnection(props);
            List<Participant> participants = new List<Participant>();
            using (var command = conn.CreateCommand())
            {
                command.CommandText = "select id_partipant from Results where id_referee=@ref";
                IDbDataParameter paramRef = command.CreateParameter();
                paramRef.ParameterName = "@ref";
                paramRef.Value = referee.Id;
                command.Parameters.Add(paramRef);

                using (var dataR = command.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idP = dataR.GetInt32(0);
                        Participant participant = participantRepository.FindOne(idP);
                        log.InfoFormat("Existing getEvaluatedParticipants with value {0}",referee);
                        participants.Add(participant);
                    }
                }
            }
            log.InfoFormat("Existing getEvaluatedParticipants with value {0}",referee);
            return participants;
        }
    }
}

File: Second Year/2nd Semester/Medii de Proiectare si Programare/Java/Laborator_1/src/main/java/repository/ResultDBRepository.java
package repository;

import domain.Participant;
import domain.Referee;
import domain.Result;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ResultDBRepository implements ResultRepository{
    private JdbcUtils dbUtils;

    private RefereeRepository refereeRepository;
    private ParticipantRepository participantRepository;

    private static final Logger logger = LogManager.getLogger();

    public ResultDBRepository(Properties props, RefereeDBRepository refereeDBRepository1, ParticipantDBRepository participantDBRepository1){
        logger.info("Initializing RefereeDBRepository with propertiesL {} ",props);
        dbUtils=new JdbcUtils(props);
        refereeRepository = refereeDBRepository1;
        participantRepository = participantDBRepository1;
    }

    @Override
    public Result findOne(Integer integer) throws IllegalArgumentException {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        Result result = null;
        try(PreparedStatement preSmt= conn.prepareStatement("select * from Results where id=?")){
            preSmt.setInt(1,integer);
            try(ResultSet resultSet = preSmt.executeQuery()){
                if(!resultSet.next())
                    return null;

                int id = resultSet.getInt("id"); // id-ul resultului
                int id_referee = resultSet.getInt("id_referee");
                Referee referee = this.refereeRepository.findOne(id_referee);
                int id_participant = resultSet.getInt("id_participant");
                Participant participant = this.participantRepository.findOne(id_participant);

                String activity = resultSet.getString("activity");
                int points = resultSet.getInt("points");

                result = new Result(id, participant,referee,activity,points);
            }
        }catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(result);
        return result;
    }

    @Override
    public Iterable<Result> getAll() {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Result> resultList = new ArrayList<>();
        try(PreparedStatement preSmt=con.prepareStatement("select * from Results")){
            try(ResultSet resultSet = preSmt.executeQuery()){
                while(resultSet.next()){
                    int id = resultSet.getInt("id");
                    int id_ref = resultSet.getInt("id_referee");
                    Referee referee = this.refereeRepository.findOne(id_ref);
                    int id_part = resultSet.getInt("id_participant");
                    Participant participant = this.participantRepository.findOne(id_part);
                    int points = resultSet.getInt("points");
                    String activity = resultSet.getString("activity");
                    Result result = new Result(id,participant,referee,activity,points);
                    resultList.add(result);
                }
            }
        } catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(resultList);
        return resultList;
    }

    @Override
    public void clear() {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        try(PreparedStatement preSmt= conn.prepareStatement("delete from Results")){
            preSmt.executeUpdate();
        }catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit();
    }

    @Override
    public void save(Result entity) {
        logger.traceEntry("saving task{} ", entity);
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into Results (id_referee,id_participant,points,activity) values (?,?,?,?)")){
            preStmt.setInt(1,entity.getReferee().getId());
            preStmt.setInt(2,entity.getParticipant().getId());
            preStmt.setInt(3,entity.getPoints());
            preStmt.setString(4,entity.getActivity());
            int result=preStmt.executeUpdate();
            logger.trace("Saved {} instances", result);
            // actualizam si punctajul participantului
            int new_points = entity.getParticipant().getPoints() + entity.getPoints();
            Participant new_participant = new Participant(entity.getParticipant().getId(),entity.getParticipant().getFirst_name(),entity.getParticipant().getLast_name(),new_points);
            this.participantRepository.update(new_participant);

        } catch(SQLException ex){
            logger.error(ex);
            System.err.println("Error DB " + ex);
        }
        logger.traceExit();
    }

    @Override
    public void update(Result entity) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        int old_points = this.participantRepository.findOne(entity.getParticipant().getId()).getPoints(); // puntajul vechi
        try(PreparedStatement preSmt= conn.prepareStatement("update Results set id_referee = ?, id_participant = ?, points = ?, activity = ? where id=?")){
            preSmt.setInt(1,entity.getReferee().getId());
            preSmt.setInt(2,entity.getParticipant().getId());
            preSmt.setInt(3,entity.getPoints());
            preSmt.setString(4,entity.getActivity());
            preSmt.setInt(5,entity.getId());
            preSmt.executeUpdate();
            // actualizam si punctajul participantului (scadem punctajul vechi, adunam punctajul nou)
            int new_points = entity.getParticipant().getPoints() - old_points + entity.getPoints();
            Participant new_participant = new Participant(entity.getParticipant().getId(),entity.getParticipant().getFirst_name(),entity.getParticipant().getLast_name(),new_points);
            this.participantRepository.update(new_participant);
        }catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(entity);
    }

    @Override
    public void delete(Integer id) {
        logger.traceEntry();
        Connection conn = dbUtils.getConnection();
        try(PreparedStatement preSmt= conn.prepareStatement("delete from Results where id=?")){
            preSmt.setInt(1,id);
            preSmt.executeUpdate();
            Result entity = findOne(id);
            // actualizam si punctajul participantului (scadem punctajul din result)
            int new_points = entity.getParticipant().getPoints() -  entity.getPoints();
            Participant new_participant = new Participant(entity.getParticipant().getId(),entity.getParticipant().getFirst_name(),entity.getParticipant().getLast_name(),new_points);
            this.participantRepository.update(new_participant);
        }catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(id);
    }


    @Override
    public List<Participant> getEvaluatedParticipants(Referee ref) {
        logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Participant> participantList = new ArrayList<>();
        try(PreparedStatement preSmt=con.prepareStatement("select * from Results")){
            try(ResultSet resultSet = preSmt.executeQuery()){
                while(resultSet.next()){
                    int id_ref = resultSet.getInt("id_referee");
                    int id_part = resultSet.getInt("id_participant");
                    Participant participant = this.participantRepository.findOne(id_part);
                    if(id_ref == ref.getId())
                        if(!participantList.contains(participant))
                            participantList.add(participant);
                }
            }
        } catch(SQLException e){
            logger.error(e);
            System.err.println("Error DB "+e);
        }
        logger.traceExit(participantList);
        return participantList;
    }
}
